src/code/z_game_dlftbls.c

```diff
@@ -1,18 +1,35 @@
 #include "global.h"
 
-#define GAMESTATE_OVERLAY(name, init, destroy, size)                                                     \
-    {                                                                                                    \
-        NULL, (uintptr_t)_ovl_##name##SegmentRomStart, (uintptr_t)_ovl_##name##SegmentRomEnd,            \
-            _ovl_##name##SegmentStart, _ovl_##name##SegmentEnd, NULL, init, destroy, NULL, NULL, 0, size \
-    }
-#define GAMESTATE_OVERLAY_INTERNAL(init, destroy, size) \
-    { NULL, 0, 0, NULL, NULL, NULL, init, destroy, NULL, NULL, 0, size }
+// Linker symbol declarations (used in the table below)
+#define DEFINE_GAMESTATE(typeName, enumName, name) DECLARE_OVERLAY_SEGMENT(name)
+#define DEFINE_GAMESTATE_INTERNAL(typeName, enumName)
+
+#include "tables/gamestate_table.h"
+
+#undef DEFINE_GAMESTATE
+#undef DEFINE_GAMESTATE_INTERNAL
+
+// Gamestate Overlay Table definition
+#define DEFINE_GAMESTATE_INTERNAL(typeName, enumName) \
+    { NULL, 0, 0, NULL, NULL, NULL, typeName##_Init, typeName##_Destroy, NULL, NULL, 0, sizeof(typeName##State) },
+
+#define DEFINE_GAMESTATE(typeName, enumName, name) \
+    { NULL,                                        \
+      (uintptr_t)_ovl_##name##SegmentRomStart,     \
+      (uintptr_t)_ovl_##name##SegmentRomEnd,       \
+      _ovl_##name##SegmentStart,                   \
+      _ovl_##name##SegmentEnd,                     \
+      NULL,                                        \
+      typeName##_Init,                             \
```

Comment:
It's probably better to leave it for a different PR as it might need some discussion, but I just realized if we're doing this we could also change the `SET_NEXT_GAMESTATE` macro to only take the type name and build the _Init and State names on its own